This patch was fed back to the upstream community:
http://code.google.com/p/libarchive/issues/detail?id=286
Given the response "I applied your patch to zip writer."
it should no longer be needed with the next release.

--- libarchive/libarchive/archive_write_set_format_zip.c.orig	Mon Mar 26 19:49:00 2012
+++ libarchive/libarchive/archive_write_set_format_zip.c	Tue Nov 27 17:31:42 2012
@@ -842,6 +842,8 @@
 	type = archive_entry_filetype(entry);
 	path = archive_entry_pathname(entry);
 
+	if (path == NULL)
+		return (0);
 	if ((type == AE_IFDIR) & (path[strlen(path) - 1] != '/')) {
 		return strlen(path) + 1;
 	} else {
